File: packages/java.com/spring-boot/src/main/java/com/proxy/reverse/Controller.java
package com.proxy.reverse;

import org.springframework.http.*;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.client.RestTemplate;

import java.util.Collections;

@RestController
@RequestMapping("/api")
public class Controller {

    private final RestTemplate restTemplate = new RestTemplate();

    // ✅ Proxy GET request to target URL
    @GetMapping
    public ResponseEntity<?> proxyGet(@RequestParam String url) {
        HttpHeaders headers = new HttpHeaders();
        headers.setAccept(Collections.singletonList(MediaType.ALL));

        HttpEntity<String> entity = new HttpEntity<>(headers);

        ResponseEntity<byte[]> response = restTemplate.exchange(
                url,
                HttpMethod.GET,
                entity,
                byte[].class
        );

        HttpHeaders responseHeaders = new HttpHeaders();
        responseHeaders.setContentType(response.getHeaders().getContentType());

        return new ResponseEntity<>(response.getBody(), responseHeaders, response.getStatusCode());
    }

    // ✅ Proxy POST request to target URL with JSON body
    @PostMapping
    public ResponseEntity<?> proxyPost(@RequestParam String url, @RequestBody String body) {
        HttpHeaders headers = new HttpHeaders();
        headers.setContentType(MediaType.APPLICATION_JSON);

        HttpEntity<String> entity = new HttpEntity<>(body, headers);

        ResponseEntity<byte[]> response = restTemplate.exchange(
                url,
                HttpMethod.POST,
                entity,
                byte[].class
        );

        HttpHeaders responseHeaders = new HttpHeaders();
        responseHeaders.setContentType(response.getHeaders().getContentType());

        return new ResponseEntity<>(response.getBody(), responseHeaders, response.getStatusCode());
    }
}

File: languages/jvm/kotlinlang.org/spring.io/src/main/kotlin/com/proxy/reverse/controller/Controller.kt
package com.proxy.reverse.controller

import org.springframework.http.*
import org.springframework.web.bind.annotation.*
import org.springframework.web.reactive.function.client.*
import org.springframework.http.server.reactive.ServerHttpRequest
import reactor.core.publisher.Mono

@RestController
class Controller(
    private val webClientBuilder: WebClient.Builder
) {

    private val webClient: WebClient by lazy {
        webClientBuilder.build() // no baseUrl; dynamic per request
    }

    @RequestMapping("/proxy", method = [RequestMethod.GET, RequestMethod.POST, RequestMethod.PUT, RequestMethod.DELETE, RequestMethod.PATCH])
    fun proxy(
        @RequestParam("url") targetUrl: String,
        request: ServerHttpRequest,
        @RequestBody(required = false) body: Mono<String>
    ): Mono<ResponseEntity<String>> {
        val method = request.method ?: HttpMethod.GET

        return webClient
            .method(method)
            .uri(targetUrl)
            .body(body, String::class.java)
            .retrieve()
            .toEntity(String::class.java)
    }
}
